Objects/weakrefobject.c

```diff
@@ -46,38 +46,59 @@ new_weakref(PyObject *ob, PyObject *callback)
 }
 
 
-/* This function clears the passed-in reference and removes it from the
- * list of weak references for the referent.  This is the only code that
- * removes an item from the doubly-linked list of weak references for an
- * object; it is also responsible for clearing the callback slot.
+/* This function is responsible for clearing the callback slot. When using this
+ * function it must be called before calling remove_weakref_from_referent,
+ * otherwise a segmentation fault will occur at build time.
  */
 static void
 clear_weakref(PyWeakReference *self)
 {
     PyObject *callback = self->wr_callback;
 
-    if (self->wr_object != Py_None) {
-        PyWeakReference **list = GET_WEAKREFS_LISTPTR(self->wr_object);
-
-        if (*list == self)
-            /* If 'self' is the end of the list (and thus self->wr_next == NULL)
-               then the weakref list itself (and thus the value of *list) will
-               end up being set to NULL. */
-            *list = self->wr_next;
-        self->wr_object = Py_None;
```

Comment:
Clearing `wr_object` should still happen in `clear_weakref()`, not `remove_weakref()`.

Reply:
The thing is, the program segfaults when I reverse the order of remove before clear

```c
weaklist_remove_weakref(list, *list);
clear_weakref(*list);
```

causes a segfault for me.